node/src/Unison/Runtime/KeyValueStore.hs

```diff
@@ -1,133 +1,50 @@
-  } deriving (Typeable)
-$(deriveSafeCopy 0 'base ''KeyValue)
-
-insertKey :: KeyHash ->  (Key, Value) -> Update KeyValue ()
-insertKey keyHash (key, value)
-    = do KeyValue store wc <- get
-         put (KeyValue (Map.insert keyHash (key, value) store) (wc + 1))
-
-deleteKey :: KeyHash -> Update KeyValue ()
-deleteKey keyHash = do
-  KeyValue store wc <- get
-  let newStore = Map.delete keyHash store
-  put (KeyValue newStore (wc + 1))
-
-lookupKey :: KeyHash -> Query KeyValue (Maybe (Key, Value))
-lookupKey key
-    = do KeyValue store dc <- ask
-         return (Map.lookup key store)
-
-lookupGT_ :: KeyHash -> Query KeyValue (Maybe (KeyHash, (Key, Value)))
-lookupGT_ keyHash = do
-  KeyValue store dc <- ask
-  pure $ Map.lookupGT keyHash store
-
-getWriteCount :: Query KeyValue Int
-getWriteCount = do
-  KeyValue store dc <- ask
-  pure dc
-
-resetWriteCount :: Update KeyValue ()
-resetWriteCount = do
-  KeyValue store wc <- get
-  put (KeyValue store 0)
-
-$(makeAcidic ''KeyValue ['insertKey, 'deleteKey, 'lookupKey, 'lookupGT_, 'getWriteCount, 'resetWriteCount])
-
-data Db = Db { acidState :: AcidState KeyValue, uid :: Hash }
-
-empty :: IO Db
-empty = do
-  g <- getStdGen
-  let (rHash, _) = random g :: (Hash, StdGen)
-      fName = Text.unpack $ Hash.base64 rHash
-  acidState <- openLocalStateFrom fName $ KeyValue Map.empty 0
-  pure $ Db acidState rHash
-
--- if acidstore doesn't exist, one is created
-load :: Hash -> IO Db
-load rHash = do
-  let fName = Text.unpack $ Hash.base64 rHash
-  acidState <- openLocalStateFrom fName $ KeyValue Map.empty 0
-  pure $ Db acidState rHash
+data Db = Db (JM.JournaledMap KeyHash (Key, Value)) Identifier
 
-close :: Db -> IO ()
-close db = closeAcidState $ acidState db
+idToText :: Identifier -> Text
+idToText (BS.Series a, BS.Series b) = decodeUtf8 $ append (Base64.encode a)
+  (Base64.encode b)
 
-cleanup :: Db -> IO ()
-cleanup db = do
-  close db
-  Directory.removeDirectoryRecursive . Text.unpack . Hash.base64 . uid $ db
+textToId :: Text -> Identifier
+textToId = undefined encodeUtf8 -- TODO
 
-maybeCollectGarbage :: Db -> IO ()
-maybeCollectGarbage (Db acidState uid) = do
-  wc <- query acidState GetWriteCount
-  case wc of
-    wc | wc >= cleanupThreshold -> do
-      let storeDirectory = Text.unpack . Hash.base64 $ uid
-      hasArchive <- Directory.doesDirectoryExist $ storeDirectory </> "Archive"
-      if hasArchive then Directory.removeDirectoryRecursive $ storeDirectory </> "Archive"
-        else pure ()
-      createArchive acidState
-      createCheckpoint acidState
-      update acidState ResetWriteCount
-    _ -> pure ()
+load :: BS.BlockStore Address -> Identifier -> IO Db
+load bs (cp, ud) = do
+  jm <- JM.fromSeries bs cp ud
+  pure $ Db jm (cp, ud)
 
 insert :: KeyHash -> (Key, Value) -> Db -> IO ()
-insert kh kv db@(Db acidState _) = do
-  update acidState $ InsertKey kh kv
-  maybeCollectGarbage db
+insert kh kv (Db journaledMap _) = JM.insert kh kv journaledMap
```

Comment:
Note that the Index methods in the Unison language are all made synchronous at this point, because I didn't introduce an STM type to Unison. The ResourcePool still flushes on release though, for future proofing.